core/src/main/java/io/grpc/internal/SpiffeUtil2.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2024 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.CertificateParsingException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Will be merged with SpiffeUtil.
+ * Provides utilities to load, extract, and parse SPIFFE ID according the SPIFFE ID standard.
+ *
+ */
+public final class SpiffeUtil2 {
+
+  private static final Logger log = Logger.getLogger(SpiffeUtil2.class.getName());
+
+  private static final Integer URI_SAN_TYPE = 6;
+  private static final String USE_PARAMETER_VALUE = "x509-svid";
+
+  /**
+   * Parses a leaf certificate from the chain to extract unique SPIFFE ID. In case of success,
+   * returns parsed TrustDomain and Path.
+   *
+   * @param certChain certificate chain to extract SPIFFE ID from
+   */
+  public static Optional<SpiffeId> extractSpiffeId(X509Certificate[] certChain)
+      throws CertificateParsingException {
+    checkArgument(checkNotNull(certChain, "certChain").length > 0, "CertChain can't be empty");
+    Collection<List<?>> subjectAltNames = certChain[0].getSubjectAlternativeNames();
+    if (subjectAltNames != null) {
+      boolean spiffeFound = false;
```

Comment:
WDYT of storing the SPIFFE ID instead of a boolean, so we don't need to iterate over the list twice?

For example,
```
Optional<String> uriSan = Optional.absent();
for (List<?> altName : subjectAltNames) {
  if (altName.length > 0 && altName.get(0).equals(URI_SAN_TYPE) {
    if (uriSan.absent()) {
      uriSan = Optional.of(altName.get(0));
    } else {
      throw exception
    }
  }
}
```

Reply:
If you add SPIFFE parsing logic to the same loop, then the following might occur:

1. SPIFFE found
2. SPIFFE parsing failed
3. Exception about SPIFFE parsing thrown
But if there is another SPIFFE, then we want to throw `Duplicate SPIFFE` exception, but not the parsing one.

So the second loop is needed for your approach as well (if we store it in the first loop, we still need to iterate over collection one more time to 're-find '), unless we want to play with indices.

What I like about the current approach is the readability - it clearly follows the spec 

1. check if there are duplicates
2. if none, do the parsing

WDYT?

Reply:
Changed as discussed offline

Reply:
Ok to resolve.